ocaml/xapi/xapi_pbd.ml

```diff
@@ -116,18 +116,22 @@ let check_sharing_constraint ~__context ~sr =
 let resync_cluster_stack_for_sr_type ~__context ~sr_sm_type =
   let required_cluster_stacks = Xapi_clustering.get_required_cluster_stacks ~__context ~sr_sm_type in
   (* This is empty if the SR requires no cluster stack *)
-  let required_clusters =
-    Db.Cluster.get_all_records ~__context
-    |> List.filter (function (cluster_ref, cluster_rec) -> List.mem cluster_rec.API.cluster_cluster_stack required_cluster_stacks)
+  let match_singleton_list = function
+    | [] -> raise Api_errors.(Server_error (no_compatible_cluster_host, []))
+    | [((_ref, _rec) as result)] -> result
+    | _ -> raise Api_errors.(Server_error (cluster_does_not_have_one_node, []))
   in
-  (* XXX For now, we only support one cluster, when we add support for
-     multiple clusters, we may want to change this behaviour. *)
-  required_clusters
-  |> List.iter
-    (fun (cluster_ref, cluster_rec) ->
-       Helpers.call_api_functions ~__context (fun rpc session_id ->
-           Client.Client.Cluster.pool_resync ~rpc ~session_id ~self:cluster_ref)
-    )
+  Db.Cluster.get_all_records ~__context
+  |> List.filter (fun (cluster_ref, cluster_rec) -> List.mem cluster_rec.API.cluster_cluster_stack required_cluster_stacks)
+  (* Only one matching cluster should be found *)
+  |>  match_singleton_list
+  |> fun (cluster_ref, cluster_rec) -> begin
+      Db.Cluster_host.get_records_where ~__context
+        ~expr:Db_filter_types.(Eq (Field "cluster", Literal (Ref.string_of cluster_ref)))
+      |> match_singleton_list
```

Comment:
I think this will fail here, because you can have multiple cluster hosts for a single cluster.
It would be a lot easier if you changed this function to also take a host parameter (the caller has the host)